File: ws01_plumbing/src/cpp03_action/src/demo02_action_client.cpp
/*
需求:编写动作客户端,可以发送一个整型数据到服务端,并处理服务端的连续反馈、最终响应结果
流程:
    前提:可以解析终端动态传入的参数
    3.自定义节点类
        3-1.创建动作客户端
        3-2.发送请求
        3-3.处理关于目标值的服务端响应(回调函数)
        3-4.处理连续反馈(回调函数)
        3-5.处理最终响应结果(回调函数)
*/

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "base_interfaces_demo/action/progress.hpp"

using base_interfaces_demo::action::Progress;
using namespace std::chrono_literals;
using std::placeholders::_1;
using std::placeholders::_2;

class ProgressActionClient : public rclcpp::Node
{
public:
    ProgressActionClient() : Node("progress_action_client_node_cpp")
    {
        // 3-1.创建动作客户端
        client_ = rclcpp_action::create_client<Progress>(this, "get_sum");
        /*
        template<typename ActionT, typename NodeT>
        typename Client<ActionT>::SharedPtr
        create_client(
        NodeT node,
        const std::string & name,
        ...)
        */

        RCLCPP_INFO(this->get_logger(), "action客户端端已创建");
    }

    // 3-2.发送请求
    void send_goal(int num)
    {
        // 3-2-1.需要连接到服务端
        if (!client_->wait_for_action_server(10s))
        {
            RCLCPP_ERROR(this->get_logger(), "服务连接失败");
            return;
        }

        // 3-2-2.发送具体请求
        auto goal = Progress::Goal();
        goal.num = num;
        rclcpp_action::Client<Progress>::SendGoalOptions options;
        options.goal_response_callback = std::bind(&ProgressActionClient::goal_response_callback, this, _1);
        options.feedback_callback = std::bind(&ProgressActionClient::feedback_callback, this, _1, _2);
        options.result_callback = std::bind(&ProgressActionClient::result_callback, this, _1);

        auto future = client_->async_send_goal(goal, options);
        /*
        std::shared_future<typename GoalHandle::SharedPtr>
        async_send_goal(const Goal & goal, const SendGoalOptions & options = SendGoalOptions())
        */
    }

    // 3-3.处理关于目标值的服务端响应
    void goal_response_callback(rclcpp_action::ClientGoalHandle<Progress>::SharedPtr goal_handle)
    {
        // std::function<void (typename ClientGoalHandle<ActionT>::SharedPtr)>;
        if (!goal_handle)
        {
            RCLCPP_INFO(this->get_logger(), "目标请求被服务端拒绝");
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "目标处理中...");
        }
    }

    // 3-4.处理连续反馈
    void feedback_callback(rclcpp_action::ClientGoalHandle<Progress>::SharedPtr goal_handle,
                           const std::shared_ptr<const Progress::Feedback> feedback)
    {
        /*
        std::function<void (
            typename ClientGoalHandle<ActionT>::SharedPtr,
            const std::shared_ptr<const Feedback>)>;
        */
        (void)goal_handle;
        int progress = (int)(feedback->progress * 100);
        RCLCPP_INFO(this->get_logger(), "当前进度:%d%%", progress);
    }

    // 3-5.处理最终响应
    void result_callback(const rclcpp_action::ClientGoalHandle<Progress>::WrappedResult &result)
    {
        // std::function<void (const WrappedResult & result)>
        // 通过状态码判断最终结果状态
        if (result.code == rclcpp_action::ResultCode::SUCCEEDED)
        {
            RCLCPP_INFO(this->get_logger(), "最终结果为:%d", result.result->sum);
        }
        else if (result.code == rclcpp_action::ResultCode::ABORTED)
        {
            RCLCPP_INFO(this->get_logger(), "被中断!");
        }
        else if (result.code == rclcpp_action::ResultCode::CANCELED)
        {
            RCLCPP_INFO(this->get_logger(), "被取消!");
        }
        else
        {
            RCLCPP_INFO(this->get_logger(), "出现未知异常!");
        }
    }

private:
    rclcpp_action::Client<Progress>::SharedPtr client_;
};

int main(int argc, char const *argv[])
{
    if (argc != 2)
    {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "请提交一个整型数据");
        return 1;
    }

    rclcpp::init(argc, argv);

    auto node = std::make_shared<ProgressActionClient>();
    node->send_goal(atoi(argv[1]));
    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

File: ws01_plumbing/src/py03_action/py03_action/demo02_action_client.py
"""
需求:编写动作客户端,可以发送一个整型数据到服务端,并处理服务端的连续反馈、最终响应结果
流程:
    前提:可以解析终端动态传入的参数
    3.自定义节点类
        3-1.创建动作客户端
        3-2.发送请求
        3-3.处理关于目标值的服务端响应(回调函数)
        3-4.处理连续反馈(回调函数)
        3-5.处理最终响应结果(回调函数)
"""

import sys

import rclpy
from rclpy.logging import get_logger
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle
from base_interfaces_demo.action import Progress


class ProgressActionClient(Node):
    def __init__(self):
        super().__init__("progress_action_client_node_py")
        self.get_logger().info("动作通信客户端已创建")

        # 3-1.创建动作客户端
        self.client = ActionClient(self, Progress, "get_sum")

    # 3-2.发送请求
    def send_goal(self, goal_num: int):
        pass
        # 连接服务端
        self.client.wait_for_server(10)

        # 发送请求
        goal = Progress.Goal()
        goal.num = goal_num
        self.future = self.client.send_goal_async(goal, self.fb_callabck)
        self.future.add_done_callback(self.goal_response_callback)

    # 3-3.处理关于目标值的服务端响应
    def goal_response_callback(self, future):
        # 获取目标句柄
        goal_handle: ClientGoalHandle = future.result()

        # 判断目标是否被正常接收
        if not goal_handle.accepted:
            self.get_logger().error("目标被拒绝了")
            return
        self.get_logger().info("目标被接收,正在处理中...")

        # 3-5.处理最终响应结果
        self.result_future = goal_handle.get_result_async()
        self.result_future.add_done_callback(self.get_result_callback)

    # 3-4.处理连续反馈
    def fb_callabck(self, fb_msg):
        progress = fb_msg.feedback.progress
        self.get_logger().info(f"连续反馈数据{progress*100:.1f}%")

    # 3-5.处理最终响应结果
    def get_result_callback(self, future):
        result = future.result().result
        self.get_logger().info(f"最终结果:{result.sum}")


def main():
    # 动态解析传入的参数
    if len(sys.argv) != 2:
        get_logger("rclpy").error("请提交一个整型数据")
        return

    rclpy.init()
    action_client = ProgressActionClient()
    action_client.send_goal(int(sys.argv[1]))
    rclpy.spin(action_client)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
